packages/validator/src/services/attestation.ts

```diff
@@ -94,6 +94,16 @@ export class AttestationService {
           resolve();
         }
       };
+      // The waitForBlockSlot should execute once per slot
+      // And per slot there should be a ValidatorEvent.chainHead event emitted
+      // In the unexpected situation where ValidatorEvent.chainHead is never emitted
+      // or is emitted but takes longer to propagate to the validator,
+      // to prevent accumulating listeners for ValidatorEvent.chainHead and triggering
+      // MaxListenersExceededWarning clear listener for ValidatorEvent.chainHead
+      // if any before setting a new one.
+      if (this.emitter.listenerCount(ValidatorEvent.chainHead) > 0) {
+        this.emitter.off(ValidatorEvent.chainHead, headListener);
+      }
       this.emitter.on(ValidatorEvent.chainHead, headListener);
```

Comment:
This solution looks very strange, I don't follow the argument. The handler `headListener` is created in this scope so it can be proven that `headListener` has never been subscribed to this event, it's impossible. So the line `this.emitter.off(ValidatorEvent.chainHead, headListener)` is useless.

I also want see first a hard proof that the MaxListenersExceededWarning comes actually from this specific listener, and not from sleep()

Reply:
> The handler headListener is created in this scope so it can be proven that headListener has never been subscribed to this event, it's impossible

In the case where `waitForBlockSlot` is called multiple times (before the unsubscribe), it is possible to have `headListener` subscribed multiple times. The intention of the check is to prevent this. But I will look into this again to ascertain your concern.

> I also want see first a hard proof that the MaxListenersExceededWarning comes actually from this specific listener, and not from sleep()

I am almost certain it does not come from the `sleep` as no where in the `sleep` code path is `EventEmitter` used. The `sleep` code path makes use of `AbortSignal` and from what I see, that uses a different API. But will try and reproduce and confirm this.